README.md
NodeBase: A node base class for Javascript and Coffeescript (logging, options, defaults and EventEmitter)
============================================

This module is the mother of all Objects of my Projects. It adds logging facility (with log levels, colored output and fancy stacktrace extraction of the current function you are in) that can be turned on and off and EventEmitter support. It provides with standard options and also provide some utility functions (namely now, merge, uuid and node_ver). Feel free to fork this and provide your own implementation. This ships as a coffee version and a JavaScript version. 

NOTE: As of version 0.5.0 development is just done in Coffeescript!


## Why?

a) Cause I find it a cool idea, to have a base class, which solves all the recurring tasks a class must handle in day to day business.

b) In nearly all node projects classes inherit often from the EventEmitter Object. However this is not enough base functionality. You can use this class everywhere you would use EventEmitter and get fancy logging, options and defaults for free!

## How to use

To install and run do the following

	git clone https://github.com/dotmaster/nodeBase
	
or

  npm install nodeBase

### Implementing it on your project

In Javascript: 

     var nodeBase = require('path/to/nodeBase'),
     util = require ('util');
     
     util.inherits(someClass, nodeBase);
     function someClass(opts){
       this.defaults={
          put:'someDefaultsHere'
        }
       nodeBase.apply(this, arguments); //pass the opts to the parrent constructor, will set this.options as a mixin of defaults and opts
     }
     someClass.prototype.someMember = function(){
       this.error('hello there')
       this.log('hello there')
     }
 
     //then somewhere in your code
 
     var myObj = new someClass({logging:true, logLevel: 'WARN', some: opts}, {someMore: 'defaults'});
     myObj.someMember(); 
     //will output 
     //[someClass.someMember id:1] --Fri, 04 Mar 2011 11:53:16 GMT  [ERROR] hello there
     //[someClass.someMember id:1] --Fri, 04 Mar 2011 11:53:16 GMT  [LOG] hello there
                ^             ^                                       ^
                |             |                                       |
            the class   the object id                   OUTPUT OF LOGLEVEL WILL BE COLORED
        and member function  
        
In Coffeescript:

There are tow ways to implement NodeBase:
a) Inheritance

    nodeBase = require 'path/to/nodeBase/'
    util = require 'util'

    class someClass extends nodeBase
      constructor:(opts) ->
        @defaults =
          put:'someDefaultsHere'
        super(opts) #pass options to super, will set @options as a mixin of defaults and opts, no need to pass defaulkts here
      someMember: => @log 'hello there'
  
    myObj = new someClass 
      logging: true
      some:'opts',
        someMore: 'defaults'
    myObj.someMember 
    # will output 
    #[someClass.someMember id:1] --Fri, 04 Mar 2011 11:53:16 GMT  [ERROR] hello there
    #[someClass.someMember id:1] --Fri, 04 Mar 2011 11:53:16 GMT  [LOG] hello there    

b) as a mixin (beginning from version 0.6.0)
    nodeBase = require 'path/to/nodeBase/'
    util = require(if process.binding('natives').util then 'util' else 'sys')
    
    #make a static logging shortcut
    log = -> filteredClient.log arguments...
    
    #class has another base class than nodeBase already
    class filteredClient extends Client
      #add some static class defaults
      @defaults: 
        logging:true
      nodeBase.static(@); #add static @options and @defaults to class
      constructor:(opts) -> #pass in some object level options and the defaults
        #add some object level defaults
        @defaults =
          put:'someDefaultsHere'
        #must call the constructor before mixing in nodeBase! (to reduce risk of potentially shallowing constructor functions of Client)
        super(arguments...)
        #nodeBase as a mixin, using merge is better then just saying @[i]=n[i]
        @[i] = NodeBase.merge @[i] ||= {} = n[i] for i, val of n = new nodeBase(opts,@defaults) #remember to pass the defaults
      register: => #some member functions
      
      myObj = new filteredClient 
        logging: true
        logLevel: 'WARN'
        some:'opts'

      myObj.register()
            
      #will output
      #[new filteredClient id:1]  -- Mon, 07 Mar 2011 12:52:08 GMT  [ERROR] awesome!!!
      #[filteredClient.register id:1]  -- Mon, 07 Mar 2011 12:52:08 GMT  [WARN] hello there

    
## Reserved words

There are some variable and function names, that you can't use in your derived class, cause the base class is using it.
Those are: this.defaults, this.options, this.emit, this.on, this.log, this.warn, this.info, this.error, this.LOG_LEVELS, this._uuid, this._cid_, this._getTotalCids, this._addContext (despite this.constructor)

## Features

### Logging
- fancy (highly experimental) stack trace extraction
- log levels
- colored output
- as of version 0.5.0 introduced static logging (see test directory for usage)

### Auto Id and Uuid creation for objects
- every new object gets a uuid and a id (starting from 1) stored unde _uuid and _id

### Global cache of objects created (highly experimental)
- version 0.5.0 has a global object cache (with a custom capped collection implementation to avoid memory leaks)


## Options and Defaults
you can pass in the following defaults to your class:

- logging: true/false turns logging on and off (DEFAULT is false)
- logLevel: 'ALL' (same as 'LOG'), 'INFO', 'WARN', 'ERROR' turns logging on and off based on LOGLEVEL (DEFAULT is ALL)
- printLevel: true/false prints the logging level in the output like [WARN]  (DEFAULT is true)
- printContext: true/false prints the current context e.g. the class you are in (DEFAULT is true)
- useStack: true/false extracts function name from Stacktrace (highly experimental) for printing out the current context (DEFAULT is true)
- emitLog: true/false if to emit the log messages as an event of form ('level', {message: logMessage, data:anyObj}) (DEFAULT is true)
- autoId, autoUuid: true/false if to generate an Id, Uuid for each object (DEFAULT is true)
- addToCollection: true/false if to turn on the global caching (DEFAULT is false)

## Using the utility functions
  nodeBase.uuid()
    "B1BBA3DC-B1DE-477A-9CF8-7DB0000BA766"
  nodeBase.now()
    "Thu, 03 Mar 2011 22:28:39 GMT"
  nodeBase.node_ver
    major: 0
    minor: 4
    release: 1
    
## Running tests

NodeBase has been tested against version node >0.2.0 (only compiled Javascript version)

There is merely nothing right now: just type
    cd test
    //for Javascript
    node test
    //for Coffeescript
    node test.coffee
    
Furthermore there are some performance tests regarding the Cache implementation [www.perf.com/browse/Gregor]  (http://www.perf.com/browse/Gregor)

## Todo
- eventually export log functions also statically
- feel free to extend NodeBase in whatever way you want. E.g. a database logger, config reader and writer, etc.
- eventually add an option for a static objects lookup function (danger of memory leak though)


## Credits

- Gregor Schwab &lt;greg@synaptic-labs.net&gt; ([dotmaster](http://github.com/dotmaster))

## License 

(The MIT License)

Copyright (c) 2011 Gregor Schwab &lt;dev@synaptic-labs.net&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

n.n.n / 2011-03-06 
==================



n.n.n / 2011-03-05 
==================

  * added coffee-script dependency, log emit perfection
  * fixed static function bindings @log: @log=

n.n.n / 2011-03-05 
==================

  * fixed static function bindings @log: @log=

n.n.n / 2011-03-05 
==================

  * fixed static function bindings @log: @log=

n.n.n / 2011-03-04 
==================



n.n.n / 2011-03-04 
==================



n.n.n / 2011-03-04 
==================

  * added defaults for classes
  * added defaults for new classes
  * added defaults for new classes
  * added defaults for new classes
  * added defaults for new classes
  * initial commit
  * initial commit
  * initial commit
  * initial commit
  * initial commit
  * initial commit

n.n.n / 2011-03-04 
==================

  * added defaults for classes

n.n.n / 2011-03-04 
==================



n.n.n / 2011-03-03 
==================



n.n.n / 2011-03-03 
==================




test/test.js
var nodeBase = require("../index"), 
 util = require(process.binding('natives').util ? 'util' : 'sys');
 util.inherits(someClass, nodeBase);
 
 //you need to add these lines to get static logging functions working 
 someClass.defaults = { logging:true }
 nodeBase.static(someClass) //add static functions
 var log= someClass.log
 
 function someClass(){
   this.defaults={
     put:'someDefaultsHere',
     defaults:'canNotBeOverridden'
   }   
   nodeBase.apply(this, arguments);
  this.info('wow this is cool');
 }
 someClass.prototype.someMember = function(){
   this.log('hello there');
   this.info('hello there');
   this.warn('hello there');
   this.error('hello there');         
  }
 

 
 var myObj = new someClass({logging:true, logLevel:'INFO', hello:'opts'});
 myObj.someMember(); 
 
 log ('Number of Objects created ' + someClass.getTotalIds())
 //should output
 //[new someClass] --Fri, 04 Mar 2011 11:53:16 GMT  [INFO] wow this is cool <--OUTPUT OF LOGLEVEL WILL BE COLORED
 //[someClass.someMember] --Fri, 04 Mar 2011 11:53:16 GMT  [INFO] hello there
 //[someClass.someMember] --Fri, 04 Mar 2011 11:53:16 GMT  [WARN] hello there
 //[someClass.someMember] --Fri, 04 Mar 2011 11:53:16 GMT  [ERROR] hello there

test/nodeBase.js
var events = require('events');
var util = require(process.binding('natives').util ? 'util' : 'sys');
//NodeBase is an EventEmitter
util.inherits(NodeBase, events.EventEmitter);
/**
 * @desc this is the mother of all Objects; a node Base Class with (logging and options, defaults)
 */

module.exports=NodeBase;

//LOG LEVELS
var L = 0
var LL =
 { ALL: L++
 , LOG : L++
 , INFO : L++
 , WARN : L++
 , ERROR : L++
 };
for (L in LL) LL[LL[L]] = L; 
module.exports.LOG_LEVELS = LL;
   
function NodeBase(opts, defaults){
  defaults = defaults || {};
  events.EventEmitter.call(this);
	//process.EventEmitter.call(this);
	var self=this;
	//self.defaults=merge(self.defaults, defaults); //leave defaults like they are
	options({
	  //yourDefaultsGoHere: true,
	  logging: false,
	  logLevel: 'ALL',
	  printLevel: true,
	  printContext: true,
	  useStack: true,
		cacheSize: 5 //a fun property whatever this means
	}, self.defaults, defaults, opts, self);
	//loglevel
 

  this.LOG_LEVELS = LL; //make loglevel available in the object
  this._checkLogLevel = function(level){
    return (LL[this.options.logLevel] <= LL[level]);
  }
}

//ADD THE CLASSNAME AND A TIMESTAMP TO THE LOGGING OUTPUT
NodeBase.prototype._addContext = function _addC(a, level){
  var args = Array.prototype.slice.call(a);
  if(level && this.options.printLevel)  args.unshift(stylize(level)); 
  var stack;  
  try{
    if (this.options.useStack) stack = new Error().stack.split('at ')[3].match(/(.*)\s\(/)[1]; // selct everything before parenthesis
  }catch(e){  } 
  var stack = stack || this.constructor.name
  if (this.options.printContext) args.unshift('['+stack+'] ' + '--' + now()+ ' '); 
  return args;   
}
NodeBase.prototype.log = function(a){ if (this.options.logging && this._checkLogLevel('LOG')) console.log.apply(this, this._addContext(arguments, 'LOG'));}
NodeBase.prototype.warn = function(a){ if (this.options.logging && this._checkLogLevel('WARN')) console.warn.apply(this, this._addContext(arguments, 'WARN'));}
NodeBase.prototype.info = function(a){ if (this.options.logging && this._checkLogLevel('INFO')) console.info.apply(this, this._addContext(arguments, 'INFO'));}  
NodeBase.prototype.error = function(a){ if (this.options.logging && this._checkLogLevel('ERROR')) console.error.apply(this, this._addContext(arguments, 'ERROR'));}
  // we try connecting every n milli seconds. On errors n is always doubled.

function now(){
	return new Date().toUTCString();
}
NodeBase.now = module.exports.now = now;

function options(defaults){
  var args = Array.prototype.slice.call(arguments);
  var self = args[args.length-1]
  if(self instanceof NodeBase)
    args.pop(),
	  self.options = merge.apply(null, args);
  else{
    return merge.appy(null, arguments);
  }
}
NodeBase.options = module.exports.options = options;

// a mixin function similar to _.extend
function merge(obj){
  Array.prototype.slice.call(arguments, 1).forEach (function(source) {
     for (var prop in source) obj[prop] = source[prop];
   });
   return obj;
};
NodeBase.merge = module.exports.merge = merge;

//the node version
var node_ver = null;
(function(){
	if( node_ver ) return node_ver;
	var ver = process.version,
		rex = /^v(\d+)\.(\d+)\.(\d+)/i;

	var matches = ver.match(rex);
	if( !matches ){
		throw "Unable to determine node version";
	}

	node_ver = { major: (~~matches[1]), minor: (~~matches[2]), release: (~~matches[3]) };
	return node_ver;
})();
NodeBase.node_ver = module.exports.node_ver = node_ver;

//Convert arguments to array
var arrize = function(ary, from){
    return Array.prototype.slice.call(ary, from || 0);
};

//a bind function similar to _.bind
//Bind  proxy objects to function
function glue(f, obj){
	var oargs = arrize(arguments, 2);
	return function(){
		var iargs = arrize(arguments);
		return f.apply && f.apply(obj, oargs.concat(iargs));
	}
}
NodeBase.glue = module.exports.glue = glue;

/**
 * Code taken from Robert Kieffer UUID
 */

/*!
Math.uuid.js (v1.4)
http://www.broofa.com
mailto:robert@broofa.com

Copyright (c) 2010 Robert Kieffer
Dual licensed under the MIT and GPL licenses.
*/

//A UUID GENERATROR FUNCTION
 // Private array of chars to use
var CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'.split(''); 

var UUID = {};

UUID.uuid = function (len, radix) {
	var chars = CHARS, uuid = [];
	radix = radix || chars.length;

	if (len) {
		// Compact form
		for (var i = 0; i < len; i++) uuid[i] = chars[0 | Math.random()*radix];
	} else {
		// rfc4122, version 4 form
		var r;

		// rfc4122 requires these characters
		uuid[8] = uuid[13] = uuid[18] = uuid[23] = '-';
		uuid[14] = '4';

		// Fill in random data.  At i==19 set the high bits of clock sequence as
		// per rfc4122, sec. 4.1.5
		for (var i = 0; i < 36; i++) {
			if (!uuid[i]) {
				r = 0 | Math.random()*16;
				uuid[i] = chars[(i == 19) ? (r & 0x3) | 0x8 : r];
			}
		}
	}

	return uuid.join('');
};

NodeBase.uuid = module.exports.UUID = UUID;

//Stylize color helper
var stylize = function(level) {
  // http://en.wikipedia.org/wiki/ANSI_escape_code#graphics
  var styles =
      { 'bold' : [1, 22],
        'italic' : [3, 23],
        'underline' : [4, 24],
        'inverse' : [7, 27],
        'white' : [37, 39],
        'grey' : [90, 39],
        'black' : [30, 39],
        'blue' : [34, 39],
        'cyan' : [36, 39],
        'green' : [32, 39],
        'magenta' : [35, 39],
        'red' : [31, 39],
        'yellow' : [33, 39] };

  var style =
      { 'WARN': 'magenta',         
        'ERROR': 'red',              
        'INFO': 'cyan',        
        'LOG': 'green',
      }[level];

  if (style) {
    return '\033[' + styles[style][0] + 'm' + '[' + level + ']' +
           '\033[' + styles[style][1] + 'm';
  } else {
    return str;
  }
};
if (typeof global !== 'undefined' ) var colors=true;
if (! colors) {
  stylize = function(level) { return '['+ level+ '] '; };
}





test/test-coffee.js
require('coffee-script');
require('./test.coffee')


deprecated/nodeBase.js
var events = require('events');
var util = require(process.binding('natives').util ? 'util' : 'sys');
//NodeBase is an EventEmitter
util.inherits(NodeBase, events.EventEmitter);
/**
 * @desc this is the mother of all Objects; a node Base Class with (logging and options, defaults)
 */

module.exports=NodeBase;

//LOG LEVELS
var L = 0
var LL =
 { ALL: L++
 , LOG : L++
 , INFO : L++
 , WARN : L++
 , ERROR : L++
 };
for (L in LL) LL[LL[L]] = L; 
module.exports.LOG_LEVELS = LL;
   
function NodeBase(opts, defaults){
  defaults = defaults || {};
  events.EventEmitter.call(this);
	//process.EventEmitter.call(this);
	var self=this;
	//self.defaults=merge(self.defaults, defaults); //leave defaults like they are
	options({
	  //yourDefaultsGoHere: true,
	  logging: false,
	  logLevel: 'ALL',
	  printLevel: true,
	  printContext: true,
	  useStack: true,
	  emitLog: true,
	  autoUuid: true,
	  autoId: true,
		cacheSize: 5 //a fun property whatever this means
	}, self.defaults, defaults, opts, self);
	//loglevel
 

  this.LOG_LEVELS = LL; //make loglevel available in the object
  this._checkLogLevel = function(level){
    return (LL[this.options.logLevel] <= LL[level]);
  }
  
  if (this.options.autoUuid) this._uuid = NodeBase.uuid();
	if (this.options.autoId) this._id = NodeBase.cid(this);
	if (this.options.autoId) this._getTotalCids = function(){ return getTotalCids(this)} //actually this is just a counter of times the constructor was called
}

//ADD THE CLASSNAME AND A TIMESTAMP TO THE LOGGING OUTPUT
NodeBase.prototype._addContext = function _addC(a, level){
  var args = Array.prototype.slice.call(a), id="";
  var copy = args.slice(1, args.length);
  if(level && this.options.printLevel)  args.unshift(stylize(level)); 
  var stack;  
  try{
    if (this.options.useStack) stack = new Error().stack.split('at ')[3].match(/(.*)\s\(/)[1]; // selct everything before parenthesis
  }catch(e){  } 
  var stack = stack || this.constructor.name
  
  if (this.options.autoId) id= ' id:'+ this._id;
  if (this.options.printContext) args.unshift('['+stack + id + '] ' + '--' + now()+ ' '); 

  //emit a log event
  if(this.options.emitLog) this.emit(level, {'message': args.join(' ') ,'data': copy })
  
  return args;   
}
NodeBase.prototype.log = function(a){ if (this.options.logging && this._checkLogLevel('LOG')) console.log.apply(this, this._addContext(arguments, 'LOG'));}
NodeBase.prototype.warn = function(a){ if (this.options.logging && this._checkLogLevel('WARN')) console.warn.apply(this, this._addContext(arguments, 'WARN'));}
NodeBase.prototype.info = function(a){ if (this.options.logging && this._checkLogLevel('INFO')) console.info.apply(this, this._addContext(arguments, 'INFO'));}  
NodeBase.prototype.error = function(a){ if (this.options.logging && this._checkLogLevel('ERROR')) console.error.apply(this, this._addContext(arguments, 'ERROR'));}
  // we try connecting every n milli seconds. On errors n is always doubled.

function now(){
	return new Date().toUTCString();
}
NodeBase.now = module.exports.now = now;

function options(defaults){
  var args = Array.prototype.slice.call(arguments);
  var self = args[args.length-1]
  if(self instanceof NodeBase)
    args.pop(),
	  self.options = merge.apply(null, args);
  else{
    return merge.appy(null, arguments);
  }
}
NodeBase.options = module.exports.options = options;

// a mixin function similar to _.extend
function merge(obj){
  Array.prototype.slice.call(arguments, 1).forEach (function(source) {
     for (var prop in source) obj[prop] = source[prop];
   });
   return obj;
};
NodeBase.merge = module.exports.merge = merge;

//the node version
var node_ver = null;
(function(){
	if( node_ver ) return node_ver;
	var ver = process.version,
		rex = /^v(\d+)\.(\d+)\.(\d+)/i;

	var matches = ver.match(rex);
	if( !matches ){
		throw "Unable to determine node version";
	}

	node_ver = { major: (~~matches[1]), minor: (~~matches[2]), release: (~~matches[3]) };
	return node_ver;
})();
NodeBase.node_ver = module.exports.node_ver = node_ver;

//Convert arguments to array
var arrize = function(ary, from){
    return Array.prototype.slice.call(ary, from || 0);
};

//a bind function similar to _.bind
//Bind  proxy objects to function
function glue(f, obj){
	var oargs = arrize(arguments, 2);
	return function(){
		var iargs = arrize(arguments);
		return f.apply && f.apply(obj, oargs.concat(iargs));
	}
}
NodeBase.glue = module.exports.glue = glue;

/**
 * Code taken from Robert Kieffer UUID
 */

/*!
Math.uuid.js (v1.4)
http://www.broofa.com
mailto:robert@broofa.com

Copyright (c) 2010 Robert Kieffer
Dual licensed under the MIT and GPL licenses.
*/

//A UUID GENERATROR FUNCTION
 // Private array of chars to use
var CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'.split(''); 

var UUID = {};

UUID.uuid = function (len, radix) {
	var chars = CHARS, uuid = [];
	radix = radix || chars.length;

	if (len) {
		// Compact form
		for (var i = 0; i < len; i++) uuid[i] = chars[0 | Math.random()*radix];
	} else {
		// rfc4122, version 4 form
		var r;

		// rfc4122 requires these characters
		uuid[8] = uuid[13] = uuid[18] = uuid[23] = '-';
		uuid[14] = '4';

		// Fill in random data.  At i==19 set the high bits of clock sequence as
		// per rfc4122, sec. 4.1.5
		for (var i = 0; i < 36; i++) {
			if (!uuid[i]) {
				r = 0 | Math.random()*16;
				uuid[i] = chars[(i == 19) ? (r & 0x3) | 0x8 : r];
			}
		}
	}

	return uuid.join('');
};

module.exports.UUID = UUID;
NodeBase.uuid = UUID.uuid;

var cids = {};

function getTotalCids(obj){
  if(obj && obj.constructor.name)
    return cids[obj.constructor.name] || 0;
  else
    return cids['NodeBase'] || 0;
}

function cid(obj){
  //group numbering based on constructor if an object is passed and the name exists
  if(obj && obj.constructor.name)
    return ++cids[obj.constructor.name] || (cids[obj.constructor.name]= 1) //works cause NaN is false
    //cids[obj.constructor.name]=cids[obj.constructor.name]?cids[obj.constructor.name]++:1;
  //if not make a generic numbering for all objects
  else 
    return ++cids['NodeBase'] || (cids['NodeBase']= 1)
    //cids['NodeBase']=cids['NodeBase']?cids['NodeBase']++:1;
}

NodeBase.cid = module.exports.cid = cid;

//Stylize color helper
var stylize = function(level) {
  // http://en.wikipedia.org/wiki/ANSI_escape_code#graphics
  var styles =
      { 'bold' : [1, 22],
        'italic' : [3, 23],
        'underline' : [4, 24],
        'inverse' : [7, 27],
        'white' : [37, 39],
        'grey' : [90, 39],
        'black' : [30, 39],
        'blue' : [34, 39],
        'cyan' : [36, 39],
        'green' : [32, 39],
        'magenta' : [35, 39],
        'red' : [31, 39],
        'yellow' : [33, 39] };

  var style =
      { 'WARN': 'magenta',         
        'ERROR': 'red',              
        'INFO': 'cyan',        
        'LOG': 'green',
      }[level];

  if (style) {
    return '\033[' + styles[style][0] + 'm' + '[' + level + ']' +
           '\033[' + styles[style][1] + 'm';
  } else {
    return str;
  }
};
if (typeof global !== 'undefined' ) var colors=true;
if (! colors) {
  stylize = function(level) { return '['+ level+ '] '; };
}



